README.md
# Now

I wrote this first for sample code, then for myself, then for friends.

I didn't intend to push this out but darn if it isn't useful.

## Usage

```
OVERVIEW: 

Check the time at a given location, "now Sao Paolo Brazil". Locations
are diacritical and case insensitive.  Use postcodes, cities, states,
countries, even place names like "now Lincoln Memorial"

When it's this time here: "now --local 5PM Bath UK"
When it's that time there: "now --remote 5PM Bath UK"

Valid time styles: 5PM, 5:30PM, 17:30, 1730. (No spaces.)

USAGE: now [--local <local>] [--remote <remote>] [<location-info> ...]

ARGUMENTS:
  <location-info>

OPTIONS:
  -@, -l, --at, --here, --local <local>
                          When it's this local time 
  -r, --when, --there, --remote <remote>
                          When it's this remote time 
  -h, --help              Show help information.

% now sao paolo brazil
São Paulo 3:50:58 PM (GMT-3 Brasilia Standard Time)
% now --local 4PM sao paolo brazil
São Paulo 9:00:00 PM (GMT-3 Brasilia Standard Time)
% now --remote 4PM sao paolo brazil
Local 1:00:00 PM (GMT-3 Brasilia Standard Time)
```

*Note: Bug filed because help is showing `when` and not `remote` as the value for the remote time.*

## Known issues

* This can break at the edges of daylight time changes.
* Casting times (local and remote) will break when VPNs change your "location"

## Installation

* Install [homebrew](https://brew.sh).
* Install [mint](https://github.com/yonaskolb/Mint) with homebrew (`brew install mint`).
* From command line: `mint install erica/now`

## Dependencies

* [Swift-Argument-Parser](https://github.com/apple/Swift-Argument-Parser)

## Building

* Build from Xcode (there's a custom build phase that installs to /usr/local/bin, so make sure you have write access)
* Build from SPM: `swift build` in the top level directory. The built utility can be found in `.build/debug/now`. Run with `swift run`

## Thanks

*I just started this section so if you pitched in and I forgot to mention you, please let me know so I can update this!*

Darren Ford (code review), Ryan Booker ([code review and improvements](https://github.com/ryanbooker/now/blob/master/Sources/now/main.swift), and not least Paul Hudson (for living in the wrong time zone)

## Help Request

I want to add tests that will work regardless of where the utility is built and tested. (I do all tests outside of Xcode right now.) 

If you have suggestions or pointers, please let me know. Thanks!


CHANGELOG.md
# Change Log

## V 1.3

* Use AsyncParsableCommand and async versions of CoreLocation methods.
* Update swift-argument-parser.
* The code in branch `async` now requires Swift 5.5 and macOS 10.15 Catalina and up.

## V 1.2

* Rolled back Swift tools from 5.2 to 5.1 to support Mojave.

## V 1.1

* Produce JSON output on demand.

## V 1.0

* Look up time at remote location.
* Look up specified local time at remote location.
* Look up specified remote time at current location.

now/main.swift
//  Copyright © 2020 Erica Sadun. All rights reserved.

import Foundation
import ArgumentParser

/// A command that checks the time at remote locations
struct Now: ParsableCommand {
    static var configuration = CommandConfiguration(
        discussion: """
        Check the time at a given location, "now Sao Paolo Brazil". Locations
        are diacritical and case insensitive.  Use postcodes, cities, states,
        countries, even place names like "now Lincoln Memorial"

        When it's this time here: "now --local 5PM Bath UK"
        When it's that time there: "now --remote 5PM Bath UK"
        
        Valid time styles: 5PM, 5:30PM, 17:30, 1730. (No spaces.)
        """,
        version: "1.1"
    )

    @Option(
        name: [.short, .customLong("local"), .customLong("here"), .customLong("at"), .customShort("@")],
        help: "When it's this local time.")
    var localTime: String?
    
    @Option(
        name: [.short, .customLong("remote"),  .customLong("when"), .customLong("there")],
        help: "When it's this remote time.")
    var remoteTime: String?
    
    @Flag(
        name: .shortAndLong,
        help: "Output JSON results.")
    var json: Bool = false
    
    @Argument(parsing: .remaining)
    var locationInfo: [String]

    func run() throws {
        guard
            localTime == nil || remoteTime == nil
            else { throw RuntimeError.localRemoteOverlap }

        let hint = locationInfo.joined(separator: " ")
        try PlaceFinder.showTime(from: hint, at: localTime ?? remoteTime, castingTimeToLocal: remoteTime != nil, outputJSON: json)
    }    
}

Now.main()


now/PlaceFinder.swift
//  Copyright © 2020 Erica Sadun. All rights reserved.

import Foundation
import CoreLocation

typealias PlaceFindingResult = Result<[CLPlacemark], Error>

protocol DateToStringFormatter {
    func string(from date: Date) -> String
}

extension DateFormatter: DateToStringFormatter {}
extension ISO8601DateFormatter: DateToStringFormatter {}

/// A utility type to locate a place marker and print out a time with respect to that point.
enum PlaceFinder {
    /// Retrieve a placemark from a descriptive string.
    /// - Parameter hint: A free-form location indicator, such as a city name, zip code, place of interest.
    /// - Returns: A `PlaceFindingResult` of the geocoded place hint
    static func fetchPlaceMark(from hint: String) async -> PlaceFindingResult {
        do {
            let placemarks = try await CLGeocoder().geocodeAddressString(hint)
            return .success(placemarks)
        } catch {
            return .failure(error)
        }
    }
    
    /// Display a user-localized time (medium style) for a timezone described by freeform text
    ///
    /// This uses the current time to fetch the time zone abbreviation so there will be errors at the very
    /// edges of daylight changes.
    ///
    /// - Parameters:
    ///   - hint: A free-form location indicator, such as a city name, zip code, place of interest.
    ///   - date: An absolute date that will be adjusted to a timezone, localized to the user, and printed.
    ///   - localCast: Look up the remote time and cast it to the local zone
    /// - Throws: A `RuntimeError` if the target timezone cannot be interpreted.
    static func showTime(from hint: String, at timeSpecifier: String?, castingTimeToLocal localCast: Bool = false, outputJSON: Bool) async throws {
        
        // Fetch placemark
        let placemarks = try await fetchPlaceMark(from: hint).get()
        guard !placemarks.isEmpty
            else { throw RuntimeError.locationFetchFailure }
        let placemark = placemarks[0]
        
        // Extract place and zone information from placemark
        guard
            let place = placemark.name,
            let timeZone = placemark.timeZone,
            let timeZoneAbbr = placemark.timeZone?.abbreviation(for: Date()),
            let timeZoneName = placemark.timeZone?.localizedName(for: .generic, locale: .current)
            else { throw RuntimeError.timeConversionFailure }
        
        /// A pretty-printed JSON encoder
        let jsonEncoder = JSONEncoder()
        jsonEncoder.keyEncodingStrategy = .convertToSnakeCase
        jsonEncoder.outputFormatting = .prettyPrinted
        
        /// The local time zone
        let localTimeZone = Locale.autoupdatingCurrent.calendar.timeZone

        /// The default formatter
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .none
        dateFormatter.timeStyle = .medium
        dateFormatter.timeZone = localCast ? localTimeZone : timeZone
        
        /// Flexible formatter
        let formatter: DateToStringFormatter = outputJSON ? ISO8601DateFormatter() : dateFormatter
                
        func printOutput(time: String, zone: String, zoneName: String) throws {
            switch outputJSON {
            case true:
                let located = LocatedTime(place: localCast ? "Local" : place, time: time,
                                          timeZone: zone, zoneName: zoneName)
                let json = try jsonEncoder.encode(located)
                if let jsonString = String(data: json, encoding: .utf8) {
                    print(jsonString)
                } else { throw RuntimeError.jsonError }
                
            case false:
                print(#"\#(localCast ? "Local" : "\(place)") \#(time) (\#(zone) \#(zoneName))"#)
            }
        }
        
        // Current time at remote location
        guard let timeSpecifier = timeSpecifier else {
            let time = formatter.string(from: Date())
            try printOutput(time: time, zone: timeZoneAbbr, zoneName: timeZoneName)
            return
        }
        
        // Specified time
        let date = try Date.date(from: timeSpecifier)

        // Specified time at remote location
        guard localCast == true else {
            let time = formatter.string(from: date)
            try printOutput(time: time, zone: timeZoneAbbr, zoneName: timeZoneName)
            return
        }

        // Specified remote time at local location
        guard
            let localAbbreviation = localTimeZone.abbreviation(),
            let localName = localTimeZone.localizedName(for: .generic, locale: Locale.autoupdatingCurrent)
            else { throw RuntimeError.timeConversionFailure }
        let localSeconds = localTimeZone.secondsFromGMT()
        let remoteSeconds = timeZone.secondsFromGMT()
        guard let newDate = Calendar.autoupdatingCurrent.date(byAdding: .second, value: localSeconds - remoteSeconds, to: date)
            else { throw RuntimeError.timeConversionFailure }
        let time = formatter.string(from: newDate)
        try printOutput(time: time, zone: localAbbreviation, zoneName: localName)
    }
}


now/Date+Offsets.swift
//  Copyright © 2020 Erica Sadun. All rights reserved.

import Foundation

extension Date {
        
    /// Returns a reminder date from a string formatted either hour-24:minute or hour:minute-meridian.
    ///
    /// This method constructs a date from an hour and minute representation.
    /// The date is calculated from "now", moving to midnight and adding the hour and minute components.
    /// If the new date is earlier than "now", it's pushed forward 24 hours, producing the first possible
    /// instance of that hour/minute time in the future.
    ///
    /// - Parameter string: a string formatted either as "h:ma" or "H:m"
    /// - Throws: `RuntimeError`s when unable to parse the input string.
    /// - Returns: A new date, initialized to the offset of the date either today or tomorrow.
    static func date(from string: String) throws -> Date {
        // Parse date in preferred order
        let dateFormatter = DateFormatter()
        var date: Date?
        for format in ["h:ma", "ha", "H:m", "HH", "Hm", "HHmm"] {
            dateFormatter.dateFormat = format
            if let parsed = dateFormatter.date(from: string) {
                date = parsed
                break
            }
        }
        
        // Ensure date was constructed
        guard
            let componentDate = date
            else { throw RuntimeError.timeParseFailure }
                
        // Construct YMD components from now
        let calendar = Calendar.autoupdatingCurrent
        let now = Date()
        let year = calendar.component(.year, from: now)
        let month = calendar.component(.month, from: now)
        let day = calendar.component(.day, from: now)
        
        // Construct HM components from constructed date
        let hour = calendar.component(.hour, from: componentDate)
        let minute = calendar.component(.minute, from: componentDate)

        // Combine
        guard let adjustedDate = calendar.date(from: DateComponents(year: year, month: month, day: day, hour: hour, minute: minute))
            else { throw RuntimeError.timeAdjustError }
        return adjustedDate
    }
}


now/RuntimeError.swift
//  Copyright © 2020 Erica Sadun. All rights reserved.

import Foundation

/// Errors encountered while running this command
enum RuntimeError: String, Error, CustomStringConvertible {
    var description: String { rawValue }

    /// Mutual exclusion between `at` and `when`
    case localRemoteOverlap = "Cannot specify both remote and local times. Pick one."
    
    /// Time cannot be parsed
    case timeParseFailure = "Unable to parse hours and minutes from time string."
    
    /// Time components cannot be adjusted
    case timeAdjustError = "Failed to adjust time."
    
    /// Location cannot be fetched
    case locationFetchFailure = "Unable to fetch geocoded location information"
    
    /// Conversion fail
    case timeConversionFailure = "Time zone conversion failed."
    
    /// JSON fail
    case jsonError = "Unable to convert data to JSON"
}


now/Result+Utility.swift
//  Copyright © 2020 Erica Sadun. All rights reserved.

public extension Result {
    /// Initializes a `Result` from a completion handler's `(data?, error?)`.
    ///
    /// When both data and error are non-nil, `Result` first populates the
    /// `.failure` member over the `success`.
    ///
    /// - Parameters:
    ///   - data: the optional data returned via a completion handler
    ///   - error: the optional error returned via a completion handler
    init(_ data: Success?, _ error: Failure?) {
        precondition(!(data == nil && error == nil))
        switch (data, error) {
        case (_, let failure?): self = .failure(failure)
        case (let success?, _): self = .success(success)
        default:
            fatalError("Cannot initialize `Result` without success or failure")
        }
    }
}


now/Now.swift
//  Copyright © 2020 Erica Sadun. All rights reserved.

import Foundation
import ArgumentParser

/// A command that checks the time at remote locations
@main
struct Now: AsyncParsableCommand {
    static var configuration = CommandConfiguration(
        discussion: """
        Check the time at a given location, "now Sao Paolo Brazil". Locations
        are diacritical and case insensitive.  Use postcodes, cities, states,
        countries, even place names like "now Lincoln Memorial"

        When it's this time here: "now --local 5PM Bath UK"
        When it's that time there: "now --remote 5PM Bath UK"
        
        Valid time styles: 5PM, 5:30PM, 17:30, 1730. (No spaces.)
        """,
        version: "1.3"
    )

    @Option(
        name: [.short, .customLong("local"), .customLong("here"), .customLong("at"), .customShort("@")],
        help: "When it's this local time.")
    var localTime: String?
    
    @Option(
        name: [.short, .customLong("remote"),  .customLong("when"), .customLong("there")],
        help: "When it's this remote time.")
    var remoteTime: String?
    
    @Flag(
        name: .shortAndLong,
        help: "Output JSON results.")
    var json: Bool = false
    
    @Argument(parsing: .remaining)
    var locationInfo: [String]

    func run() async throws {
        guard
            CommandLine.argc > 1
            else { throw CleanExit.helpRequest() }
        
        guard
            localTime == nil || remoteTime == nil
            else { throw RuntimeError.localRemoteOverlap }

        let hint = locationInfo.joined(separator: " ")
        try await PlaceFinder.showTime(from: hint, at: localTime ?? remoteTime, castingTimeToLocal: remoteTime != nil, outputJSON: json)
    }    
}
